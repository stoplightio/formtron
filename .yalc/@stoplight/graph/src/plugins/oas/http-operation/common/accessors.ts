import unionBy = require('lodash/unionBy');

export function getOasParameters<ParamType extends { name: string; in: string }>(
  operationParameters: ParamType[] | undefined,
  pathParameters: ParamType[] | undefined
) {
  return unionBy(operationParameters, pathParameters, (parameter: ParamType) => `${parameter.name}-${parameter.in}`);
}
