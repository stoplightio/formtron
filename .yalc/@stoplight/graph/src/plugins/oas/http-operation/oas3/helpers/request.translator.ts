import { IHttpContent, IHttpRequestBody, IHttpRequestParams } from '@stoplight/types';
import { groupBy, map, mapValues, omit } from 'lodash';
import { MediaTypeObject, ParameterObject, RequestBodyObject } from 'openapi3-ts';
import { translateMediaTypeObject } from './content.translator';

function translateOas3RequestBody(requestBodyObject: RequestBodyObject): IHttpRequestBody {
  return {
    required: requestBodyObject.required,
    description: requestBodyObject.description,
    contents: map<MediaTypeObject, IHttpContent>(requestBodyObject.content, translateMediaTypeObject),
  };
}

function translateParameterObject(parameterObject: ParameterObject) {
  return {
    ...omit(parameterObject, 'in'),
    // TODO(SL-249): we are missing examples in our types, on purpose?
    // examples: parameterObject.examples,
    content: mapValues(parameterObject.content, translateMediaTypeObject),
  };
}

export function translateOas3ToRequest(
  parameters: ParameterObject[],
  requestBodyObject?: RequestBodyObject
): IHttpRequestParams {
  return {
    ...(requestBodyObject ? { body: translateOas3RequestBody(requestBodyObject) } : {}),
    ...mapValues(groupBy(parameters, 'in'), parameterObjects => map(parameterObjects, translateParameterObject)),
  };
}
