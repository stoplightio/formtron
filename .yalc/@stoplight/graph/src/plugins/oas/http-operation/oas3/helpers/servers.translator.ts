import { INodeVariable, IServer } from '@stoplight/types';
import map = require('lodash/map');
import mapValues = require('lodash/mapValues');
import { ServerObject, ServerVariableObject } from 'openapi3-ts';

export function translateToServers(servers: ServerObject[] | undefined): IServer[] {
  return map(servers, server => ({
    description: server.description,
    url: server.url,
    variables: mapValues(
      server.variables,
      (value: ServerVariableObject) =>
        ({
          default: String(value.default),
          description: String(value.default),
          enum: map(value.enum, String),
        } as INodeVariable)
    ),
  }));
}
