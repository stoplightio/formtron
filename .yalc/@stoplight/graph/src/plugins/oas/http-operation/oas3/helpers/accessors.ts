import { get, pick, values } from 'lodash';
import { OpenAPIObject, OperationObject, SecuritySchemeObject } from 'openapi3-ts';
import { uniqFlatMap } from '../../common/utils';

export function getOas3Securities(
  operation: Partial<OperationObject>,
  spec: Partial<OpenAPIObject>
): SecuritySchemeObject[] {
  const globalSchemes = uniqFlatMap(spec.security);
  const operationSchemes = uniqFlatMap(operation.security);
  return values(pick(get(spec, 'components.securitySchemes'), operation.security ? operationSchemes : globalSchemes));
}
