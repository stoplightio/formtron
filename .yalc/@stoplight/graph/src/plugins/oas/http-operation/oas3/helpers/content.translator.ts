import { HttpParamStyles, IExample, IHttpContent, IHttpEncoding, IHttpHeaderParam } from '@stoplight/types';
import { compact, mapValues, omit, union } from 'lodash';
import { EncodingPropertyObject, HeaderObject, MediaTypeObject } from 'openapi3-ts';
import { mapDict } from '../../common/utils';

function translateEncodingPropertyObject(
  encodingPropertyObject: EncodingPropertyObject,
  property: string
): IHttpEncoding {
  const acceptableStyles: Array<string | undefined> = [
    HttpParamStyles.Form,
    HttpParamStyles.SpaceDelimited,
    HttpParamStyles.PipeDelimited,
    HttpParamStyles.DeepObject,
  ];

  if (!acceptableStyles.includes(encodingPropertyObject.style)) {
    throw new Error(
      `Encoding property style: '${encodingPropertyObject.style}' is incorrect, must be one of: ${acceptableStyles}`
    );
  }

  return {
    property,
    ...encodingPropertyObject,
    // workaround for 'style' being one of the accepted HttpParamStyles
    style: encodingPropertyObject.style as any,
    mediaType: encodingPropertyObject.contentType,
    headers: mapDict(encodingPropertyObject.headers, translateHeaderObject),
  };
}

export function translateHeaderObject(headerObject: HeaderObject, name: string): IHttpHeaderParam {
  return {
    // TODO(SL-249): we are missing examples in our types, on purpose?
    // examples: parameterObject.examples,
    ...omit(headerObject, 'examples', 'example'),
    name,
    style: HttpParamStyles.Simple,
    content: mapValues(headerObject.content, translateMediaTypeObject),
  };
}

export function translateMediaTypeObject(
  { schema, example, examples, encoding }: MediaTypeObject,
  mediaType: string
): IHttpContent {
  return {
    mediaType,
    schema,
    // Note that I'm assuming all references are resolved
    examples: compact(union(Array.of(example), examples)) as IExample[],
    encodings: mapDict(encoding, translateEncodingPropertyObject),
  };
}
