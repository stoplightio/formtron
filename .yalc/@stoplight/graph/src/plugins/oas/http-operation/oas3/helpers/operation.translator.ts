import { IHttpOperation } from '@stoplight/types';
import { ParameterObject, RequestBodyObject, ResponsesObject } from 'openapi3-ts';
import { getOasParameters } from '../../common/accessors';
import { translateToTags } from '../../common/tag.translator';
import { Oas3TranslateOperationOptions } from '../../common/types';
import { getOas3Securities } from './accessors';
import { translateOas3ToRequest } from './request.translator';
import { translateToOas3Responses } from './responses.translator';
import { translateOas3ToSecurities } from './securities.translator';
import { translateToServers } from './servers.translator';

export function translateOperation(opts: Oas3TranslateOperationOptions): IHttpOperation {
  const { operation, method, path, pathObject, spec } = opts;
  return {
    id: '?http-oas3-operation-id?',
    iid: operation.operationId,
    description: operation.description,
    deprecated: operation.deprecated,
    method,
    path,
    summary: operation.summary,
    responses: translateToOas3Responses(operation.responses as ResponsesObject),
    servers: translateToServers(operation.servers || pathObject.servers || spec.servers),
    request: translateOas3ToRequest(
      getOasParameters(operation.parameters as ParameterObject[], pathObject.parameters),
      operation.requestBody as RequestBodyObject
    ),
    tags: translateToTags(operation.tags || []),
    security: translateOas3ToSecurities(getOas3Securities(operation, spec)),
  };
}
