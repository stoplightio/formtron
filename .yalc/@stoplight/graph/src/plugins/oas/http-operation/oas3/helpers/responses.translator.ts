import { IHttpResponse } from '@stoplight/types';
import { ResponseObject, ResponsesObject } from 'openapi3-ts';
import { mapDict } from '../../common/utils';
import { translateHeaderObject, translateMediaTypeObject } from './content.translator';

function translateToOas3Response(response: ResponseObject, statusCode: string): IHttpResponse {
  return {
    code: statusCode,
    description: response.description,
    headers: mapDict(response.headers, translateHeaderObject),
    contents: mapDict(response.content, translateMediaTypeObject),
  };
}

export function translateToOas3Responses(responses: ResponsesObject): IHttpResponse[] {
  return mapDict(responses, translateToOas3Response);
}
