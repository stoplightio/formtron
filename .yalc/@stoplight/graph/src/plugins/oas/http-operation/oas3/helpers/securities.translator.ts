import { SecuritySchemeObject } from 'openapi3-ts';
import { HttpSecurityScheme } from '../../common/types';

export function translateOas3ToSecurities(securities: SecuritySchemeObject[]): HttpSecurityScheme[] {
  // return compact(
  //   map(securities, security => {
  //     const common: any = pick(security, ['type', 'description']);
  //     switch (security.type) {
  //       case 'http':
  //         return {
  //           ...common,
  //           scheme: security.scheme,
  //         } as IBasicSecurityScheme;
  //       case 'apiKey':
  //         return {
  //           ...common,
  //           name: security.name,
  //           in: security.in,
  //         } as IApiKeySecurityScheme;
  //       case 'oauth2':
  //         return {
  //           ...common,
  //           flows: security.flows,
  //         } as IOauth2SecurityScheme;
  //       case 'openIdConnect':
  //         return {
  //           ...common,
  //           openIdConnectUrl: security.openIdConnectUrl,
  //         } as IOpenIdConnectSecurityScheme;
  //     }
  //     return null;
  //   })
  // );

  // TODO(SL-249): see the implementation above. I'm basically assigning one property to another.
  // Instead I think it's fine to just 'assume' these types match.
  // This may of course become WRONG at some point of course but a proper test suite should pick that up.
  // I will leave this until reviewers take a look.
  return securities as HttpSecurityScheme[];
}
