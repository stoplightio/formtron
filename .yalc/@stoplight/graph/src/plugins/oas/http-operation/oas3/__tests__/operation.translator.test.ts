jest.mock('../../common/accessors');
jest.mock('../../common/tag.translator');
jest.mock('../helpers/accessors');
jest.mock('../helpers/request.translator');
jest.mock('../helpers/responses.translator');
jest.mock('../helpers/securities.translator');
jest.mock('../helpers/servers.translator');

import { mockPassthroughImplementation } from '@stoplight/test-utils';
import { getOasParameters } from '../../common/accessors';
import { translateToTags } from '../../common/tag.translator';
import { getOas3Securities } from '../helpers/accessors';
import { translateOperation } from '../helpers/operation.translator';
import { translateOas3ToRequest } from '../helpers/request.translator';
import { translateToOas3Responses } from '../helpers/responses.translator';
import { translateOas3ToSecurities } from '../helpers/securities.translator';
import { translateToServers } from '../helpers/servers.translator';

describe('translateOperation', () => {
  beforeEach(() => {
    mockPassthroughImplementation(
      translateToServers,
      translateOas3ToSecurities,
      translateToOas3Responses,
      translateOas3ToRequest,
      translateToTags,
      getOas3Securities,
      getOasParameters
    );
  });

  test('given no tags should translate operation with empty tags array', () => {
    expect(
      translateOperation({
        method: 'method',
        path: 'path',
        operation: {
          operationId: 'opid',
          responses: {},
          deprecated: true,
          description: 'descr',
          summary: 'summary',
        },
        spec: {
          info: {
            title: 'title',
            version: '2',
          },
          openapi: '2',
          paths: [],
        },
        pathObject: {},
      })
    ).toMatchSnapshot();
  });

  test('given some tags should translate operation with those tags', () => {
    expect(
      translateOperation({
        method: 'method',
        path: 'path',
        operation: {
          operationId: 'opid',
          responses: {},
          deprecated: true,
          description: 'descr',
          summary: 'summary',
          tags: ['tag'],
        },
        spec: {
          info: {
            title: 'title',
            version: '2',
          },
          openapi: '2',
          paths: [],
        },
        pathObject: {},
      })
    ).toMatchSnapshot();
  });

  test('given operation servers should translate operation with those servers', () => {
    expect(
      translateOperation({
        method: 'method',
        path: 'path',
        operation: {
          operationId: 'opid',
          responses: {},
          deprecated: true,
          description: 'descr',
          summary: 'summary',
          servers: [{ url: 'operation/server' }],
        },
        spec: {
          info: {
            title: 'title',
            version: '2',
          },
          openapi: '2',
          paths: [],
        },
        pathObject: {},
      })
    ).toMatchSnapshot();
  });

  test('given path servers should translate operation with those servers', () => {
    expect(
      translateOperation({
        method: 'method',
        path: 'path',
        operation: {
          operationId: 'opid',
          responses: {},
          deprecated: true,
          description: 'descr',
          summary: 'summary',
        },
        spec: {
          info: {
            title: 'title',
            version: '2',
          },
          openapi: '2',
          paths: [],
        },
        pathObject: {
          servers: [{ url: 'path/server' }],
        },
      })
    ).toMatchSnapshot();
  });

  test('given spec servers should translate operation with those servers', () => {
    expect(
      translateOperation({
        method: 'method',
        path: 'path',
        operation: {
          operationId: 'opid',
          responses: {},
          deprecated: true,
          description: 'descr',
          summary: 'summary',
        },
        spec: {
          info: {
            title: 'title',
            version: '2',
          },
          openapi: '2',
          paths: [],
          servers: [{ url: 'spec/server' }],
        },
        pathObject: {},
      })
    ).toMatchSnapshot();
  });
});
