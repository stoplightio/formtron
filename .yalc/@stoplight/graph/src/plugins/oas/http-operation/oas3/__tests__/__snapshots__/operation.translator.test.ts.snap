// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateOperation given no tags should translate operation with empty tags array 1`] = `
Object {
  "deprecated": true,
  "description": "descr",
  "id": "?http-oas3-operation-id?",
  "iid": "opid",
  "method": "method",
  "path": "path",
  "request": "translateOas3ToRequest(getOasParameters(undefined, undefined), undefined)",
  "responses": "translateToOas3Responses({})",
  "security": "translateOas3ToSecurities(getOas3Securities({\\"operationId\\":\\"opid\\",\\"responses\\":{},\\"deprecated\\":true,\\"description\\":\\"descr\\",\\"summary\\":\\"summary\\"}, {\\"info\\":{\\"title\\":\\"title\\",\\"version\\":\\"2\\"},\\"openapi\\":\\"2\\",\\"paths\\":[]}))",
  "servers": "translateToServers(undefined)",
  "summary": "summary",
  "tags": "translateToTags([])",
}
`;

exports[`translateOperation given operation servers should translate operation with those servers 1`] = `
Object {
  "deprecated": true,
  "description": "descr",
  "id": "?http-oas3-operation-id?",
  "iid": "opid",
  "method": "method",
  "path": "path",
  "request": "translateOas3ToRequest(getOasParameters(undefined, undefined), undefined)",
  "responses": "translateToOas3Responses({})",
  "security": "translateOas3ToSecurities(getOas3Securities({\\"operationId\\":\\"opid\\",\\"responses\\":{},\\"deprecated\\":true,\\"description\\":\\"descr\\",\\"summary\\":\\"summary\\",\\"servers\\":[{\\"url\\":\\"operation/server\\"}]}, {\\"info\\":{\\"title\\":\\"title\\",\\"version\\":\\"2\\"},\\"openapi\\":\\"2\\",\\"paths\\":[]}))",
  "servers": "translateToServers([{\\"url\\":\\"operation/server\\"}])",
  "summary": "summary",
  "tags": "translateToTags([])",
}
`;

exports[`translateOperation given path servers should translate operation with those servers 1`] = `
Object {
  "deprecated": true,
  "description": "descr",
  "id": "?http-oas3-operation-id?",
  "iid": "opid",
  "method": "method",
  "path": "path",
  "request": "translateOas3ToRequest(getOasParameters(undefined, undefined), undefined)",
  "responses": "translateToOas3Responses({})",
  "security": "translateOas3ToSecurities(getOas3Securities({\\"operationId\\":\\"opid\\",\\"responses\\":{},\\"deprecated\\":true,\\"description\\":\\"descr\\",\\"summary\\":\\"summary\\"}, {\\"info\\":{\\"title\\":\\"title\\",\\"version\\":\\"2\\"},\\"openapi\\":\\"2\\",\\"paths\\":[]}))",
  "servers": "translateToServers([{\\"url\\":\\"path/server\\"}])",
  "summary": "summary",
  "tags": "translateToTags([])",
}
`;

exports[`translateOperation given some tags should translate operation with those tags 1`] = `
Object {
  "deprecated": true,
  "description": "descr",
  "id": "?http-oas3-operation-id?",
  "iid": "opid",
  "method": "method",
  "path": "path",
  "request": "translateOas3ToRequest(getOasParameters(undefined, undefined), undefined)",
  "responses": "translateToOas3Responses({})",
  "security": "translateOas3ToSecurities(getOas3Securities({\\"operationId\\":\\"opid\\",\\"responses\\":{},\\"deprecated\\":true,\\"description\\":\\"descr\\",\\"summary\\":\\"summary\\",\\"tags\\":[\\"tag\\"]}, {\\"info\\":{\\"title\\":\\"title\\",\\"version\\":\\"2\\"},\\"openapi\\":\\"2\\",\\"paths\\":[]}))",
  "servers": "translateToServers(undefined)",
  "summary": "summary",
  "tags": "translateToTags([\\"tag\\"])",
}
`;

exports[`translateOperation given spec servers should translate operation with those servers 1`] = `
Object {
  "deprecated": true,
  "description": "descr",
  "id": "?http-oas3-operation-id?",
  "iid": "opid",
  "method": "method",
  "path": "path",
  "request": "translateOas3ToRequest(getOasParameters(undefined, undefined), undefined)",
  "responses": "translateToOas3Responses({})",
  "security": "translateOas3ToSecurities(getOas3Securities({\\"operationId\\":\\"opid\\",\\"responses\\":{},\\"deprecated\\":true,\\"description\\":\\"descr\\",\\"summary\\":\\"summary\\"}, {\\"info\\":{\\"title\\":\\"title\\",\\"version\\":\\"2\\"},\\"openapi\\":\\"2\\",\\"paths\\":[],\\"servers\\":[{\\"url\\":\\"spec/server\\"}]}))",
  "servers": "translateToServers([{\\"url\\":\\"spec/server\\"}])",
  "summary": "summary",
  "tags": "translateToTags([])",
}
`;
